File: medium/strings/threeSum.cpp
vector<vector<int>> threeSum(vector<int>& nums){

    int n = nums.size();
    sort(nums.begin(), nums.end());
    vector<vector<int> > res;

    for(int i = 0; i < n-2; ++i){
        if(i > 0 && (nums[i] == nums[i-1]))
            continue;
        int low = i+1, high = n-1;

        while(low < high){
            int sum = nums[i]+nums[low]+nums[high];
            if(sum < 0){
                low++;
            }
            else if(sum > 0){
                high--;
            }
            else{
                // cout << nums[i] << " " << nums[low] << " " << nums[high] << endl;
                res.push_back(vector<int>{nums[i], nums[low], nums[high]});
                while(low+1 < high && nums[low] == nums[low+1]) // skipping duplicates values
                    low++;
                while(low < high-1 && nums[high] == nums[high-1])
                    high--;
                low++, high--; // skipping last duplicate

            }
        }
    }
    return res;
}

File: medium/strings/threeSum.py
def threeSum(nums):
	res = []
	n = len(nums)
	for i in range(n-2):
		if i > 0 and nums[i] == nums[i-1]:
			continue
		low = i+1
		high = n-1

		while(low < high):
			sum = nums[i] + nums[low] + nums[high]
			if sum < 0:
				low += 1
			elif sum > 0:
				high -= 1
			else:
				res.append([nums[i], nums[low], nums[high]])
				while low+1 < high and nums[low] == nums[low+1]:
					low += 1
				while low < high-1 and nums[high] == nums[high-1]:
					high -= 1
				low += 1; high -= 1
	return res			
